File: AutomationService/AutomationService/Features/Backup/Domain/Services/LetzteSicherung.cs
namespace AutomationService.Features.Backup.Domain.Services;

/// <summary>
/// Wie der letzte automatische Sicherungslauf ausgegangen ist (§7.2, #39).
///
/// Der Lauf passiert, wenn das Fenster schon zu ist — es gibt in dem Moment
/// niemanden mehr, dem man etwas sagen könnte. Deshalb wird das Ergebnis
/// aufgehoben und beim <em>nächsten</em> Start gezeigt. Ohne das wäre eine
/// Sicherung, die seit Wochen an einem umbenannten Ordner scheitert, für den
/// Anwalt nicht von einer heilen zu unterscheiden — bis er sie braucht.
/// </summary>
/// <param name="Zeitpunkt">Wann der Lauf stattfand.</param>
/// <param name="Gelungen">Ob am Ende ein Archiv im Ablageordner lag.</param>
/// <param name="Datei">Name des Archivs, wenn der Lauf gelungen ist.</param>
/// <param name="Meldung">
/// Was schiefging, in Worten, die der Anwalt lesen kann — kein Stacktrace.
/// </param>
/// <param name="FehlerQuittiert">
/// Ob die Meldung schon einmal gezeigt und weggeklickt wurde. Sonst stünde sie
/// bei jedem Start wieder da, auch wenn der Anwalt sie längst verstanden hat.
/// Ein gelungener Lauf setzt den Merker ohnehin zurück.
/// </param>
public sealed record LetzteSicherung(
    DateTime Zeitpunkt,
    bool Gelungen,
    string? Datei,
    string? Meldung,
    bool FehlerQuittiert)
{
    /// <summary>Ein Fehlschlag, der dem Anwalt noch gezeigt werden muss.</summary>
    public bool OffenerFehler => !Gelungen && !FehlerQuittiert;
}

File: Automation_App_Frontend/lib/features/backup/domain/entities/letzte_sicherung.dart
import 'package:equatable/equatable.dart';

/// Ausgang des letzten automatischen Sicherungslaufs (§7.2).
///
/// Der Lauf passiert, wenn das Fenster schon zu ist — es gibt in dem Moment
/// niemanden, dem man etwas sagen könnte. Deshalb hebt das Backend das Ergebnis
/// auf, und die App zeigt es beim nächsten Start. Ohne das wäre eine Sicherung,
/// die seit Wochen an einem umbenannten Ordner scheitert, nicht von einer
/// heilen zu unterscheiden — bis der Anwalt sie braucht.
class LetzteSicherung extends Equatable {
  final DateTime zeitpunkt;
  final bool gelungen;

  /// Name des Archivs, wenn der Lauf gelungen ist.
  final String? datei;

  /// Was schiefging, in Worten, die der Anwalt lesen kann.
  final String? meldung;

  /// Ob die Meldung schon gezeigt und weggeklickt wurde.
  final bool fehlerQuittiert;

  const LetzteSicherung({
    required this.zeitpunkt,
    required this.gelungen,
    this.datei,
    this.meldung,
    this.fehlerQuittiert = false,
  });

  /// Ein Fehlschlag, der dem Anwalt noch gezeigt werden muss.
  bool get offenerFehler => !gelungen && !fehlerQuittiert;

  factory LetzteSicherung.fromJson(
    Map<String, dynamic> json,
  ) => LetzteSicherung(
    // Der Dienst sendet den Zeitpunkt mit Zeitzonenversatz; ohne
    // `toLocal()` stünde in der Anzeige die UTC-Uhrzeit (zwei Stunden früher).
    zeitpunkt:
        DateTime.tryParse(json['zeitpunkt'] as String? ?? '')?.toLocal() ??
        DateTime.now(),
    gelungen: json['gelungen'] as bool? ?? false,
    datei: json['datei'] as String?,
    meldung: json['meldung'] as String?,
    fehlerQuittiert: json['fehlerQuittiert'] as bool? ?? false,
  );

  @override
  List<Object?> get props => [
    zeitpunkt,
    gelungen,
    datei,
    meldung,
    fehlerQuittiert,
  ];
}
